### PR TITLE
Webgl2 compatibility

### PR DESCRIPTION
**Pull request by zaszi**
This is a continuation of @mrk-its proof-of-concept, allowing `bevy_tilemap` to now also correctly run as WebAssembly when `bevy_webgl2` is used as the rendering back-end.

Compatibility was achieved by providing an additional set of shaders compatible with WASM, as well as some conditional compilation for the rendering pipeline (meaning that no feature flags are required). Examples have been untouched as providing WASM compatibility without breaking native would be non-trivial until [Target-specific features](https://github.com/rust-lang/cargo/issues/1197) make it into stable rust.

### CODE REVIEW

**Comment by joshuajbouw** [comment]
Thanks for this and your first contribution!

I am currently in the middle of quite a large overhaul that has custom shaders in mind. I think there needs to be a bit more logic added to ensure that either the WASM shaders or the regular shaders are compiled then used. Will take a look later.

**Comment by digital7-code** [comment]
> Thanks for this and your first contribution!
> 
> I am currently in the middle of quite a large overhaul that has custom shaders in mind. I think there needs to be a bit more logic added to ensure that either the WASM shaders or the regular shaders are compiled then used. Will take a look later.

When is this coming? (custom shaders thing)

**Comment by joshuajbouw** [comment]
Honestly was going to do it last weekend but... Then Valheim came along.

File: src/chunk/render.rs [hunk 1]
```diff
@@ -45,57 +45,126 @@ macro_rules! build_chunk_pipeline {
...
+                    fragment: Some(shaders.add(Shader::from_glsl(ShaderStage::Fragment, {
+                        let _wasm = false;
+                        #[cfg(target_arch = "wasm32")]
+                        let _wasm = true;
+
+                        if _wasm {
+                            include_str!("render_web/tilemap.frag")
+                        } else {
+                            include_str!("render_native/tilemap.frag")
+                        }
```
Comment (joshuajbouw) [inline src/chunk/render.rs:57]: There is a better way to handle this with `if cfg!(target_arch = "wasm32") {`

File: src/chunk/render.rs [hunk 2]
```diff
@@ -45,57 +45,126 @@ macro_rules! build_chunk_pipeline {
...
     };
 }
 
+#[cfg(not(target_arch = "wasm32"))]
 build_chunk_pipeline!(
     CHUNK_SQUARE_PIPELINE,
     2110840099625352487,
     build_chunk_square_pipeline,
-    "tilemap-square.vert"
+    "render_native/tilemap-square.vert"
```
Comment (joshuajbouw) [inline src/chunk/render.rs:70]: Since these are the "intended" target, think that I might keep these in the root folder then web in another.

**Review by joshuajbouw — COMMENTED** [review]
Generally a good start that does absolutely seem workable. I'm really considering a best course of action because a general approach that I want to do is have this library simply `bevy` native and then have it become extensible externally. I think, I will try a re-implementation of this approach with that in mind. I have ideas on how exactly that can go down that are rather straight forward, essentially exactly this, but through a sort of custom interface that allows for others to also make their own custom shaders. I imagine that would make a few people happy at some point in the future.

File: src/chunk/render.rs (hunk 1 above)
Comment (zaszi) [reply src/chunk/render.rs:57]: Yeah, this part wasn't very ideal. The compiler kept bullying me here with conditionals applied directly to a macro, but I probably just did something wrong.

**Comment by RichoDemus** [comment]
I'm really interested in using bevy_tilemap in wasm, any updates on this? :)

**Pull request closed without merging**